sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala

```diff
@@ -0,0 +1,151 @@
+    _.containsPattern(TreePattern.UNRESOLVED_TRANSPOSE)) {
+    case t @ UnresolvedTranspose(indexColumn, child, _, _) if !t.resolved =>
+
+      // Cast the index column to StringType
+      val indexColumnAsString = indexColumn match {
+        case attr: Attribute if attr.dataType.isInstanceOf[AtomicType] =>
+          Alias(Cast(attr, StringType), attr.name)()
+        case attr: Attribute =>
+          throw new IllegalArgumentException(
+            s"Index column must be of atomic type, but found: ${attr.dataType}")
+        case _ =>
+          throw new IllegalArgumentException(
+            s"Index column must be an atomic attribute")
+      }
+
+      // Cast non-index columns to the least common type
+      val nonIndexColumnsAttr = child.output.filterNot(
+        _.exprId == indexColumn.asInstanceOf[Attribute].exprId)
+      val nonIndexTypes = nonIndexColumnsAttr.map(_.dataType)
+      val commonType = leastCommonType(nonIndexTypes)
+      val nonIndexColumnsAsLCT = nonIndexColumnsAttr.map { attr =>
+        Alias(Cast(attr, commonType), attr.name)()
+      }
+
+      // Sort by index column values, and collect the casted frame
+      val allCastCols = indexColumnAsString +: nonIndexColumnsAsLCT
+      val sortedChild = Sort(
+        Seq(SortOrder(indexColumn.asInstanceOf[Attribute], Ascending)),
+        global = true,
+        child
+      )
+      val projectAllCastCols = Project(allCastCols, sortedChild)
+      val queryExecution = sparkSession.sessionState.executePlan(projectAllCastCols)
+      val fullCollectedRows = queryExecution.executedPlan.executeCollect()
+
+      if (fullCollectedRows.isEmpty) {
+        // Return a DataFrame with a single column "key" containing non-index column names
+        val keyAttr = AttributeReference("key", StringType, nullable = false)()
+        val keyValues = nonIndexColumnsAttr.map(
+          _.name).map(name => UTF8String.fromString(name))
+        val keyRows = keyValues.map(value => InternalRow(value))
+
+        LocalRelation(Seq(keyAttr), keyRows)
+      } else {
+
+        // Transpose the matrix
+        val nonIndexColumnNames = nonIndexColumnsAttr.map(_.name)
+        val nonIndexColumnDataTypes = projectAllCastCols.output.tail.map(attr => attr.dataType)
+        val transposedMatrix = transposeMatrix(
+          fullCollectedRows, nonIndexColumnNames, nonIndexColumnDataTypes)
+        val transposedInternalRows = transposedMatrix.map { row =>
+          InternalRow.fromSeq(row.toIndexedSeq)
+        }
+
+        // Construct output attributes
+        val keyAttr = AttributeReference("key", StringType, nullable = false)()
+        val transposedColumnNames = fullCollectedRows.map { row => row.getString(0) }
+        val valueAttrs = transposedColumnNames.map { value =>
+          AttributeReference(
+            value,
+            commonType
+          )()
+        }
+
+        LocalRelation(
```

Comment:
Again, we should not turn it into `LocalRelation` during analysis, as we need observability. Let's split your code here into two parts: add casts and collect results to do transpose.

Reply:
Sounds good, done.